### ISSUE TITLE
Fused exterior twisted-clover fermions incompatible with tuning 

### ISSUE THREAD

**Issue by mikeaclark** [lattice/quda#218]
(bug report from Alex)

I'm still having problems with the fused kernels and tuning, I don't understand why. Disabling tuning yields the right result.

You can actually test it: if you use the same source (not point, fill the spinor with some random stuff), you can try to run invert_test with and without tuning on multiGPU with fused kernels, and you should see at the beginning:

"Prepared source = XXXXXXXXe+YY"

The number should be the same, regardless of the tuning, but this is not the case.  I found some problems with the labeling during tuning: the application of the clover term was not labeled properly, and the dslash kernels could be labeled wrong. After correcting this in dslash_quda.cu:

- Just add:

```
    if (twist == QUDA_TWIST_GAMMA5_DIRECT) {
      writeAuxString("direct");
    } else {
      writeAuxString("inverse");
    }
```
in the constructor, and change tuneKey to

```
    TuneKey tuneKey() const { return TuneKey(in->VolString(),
typeid(*this).name(), aux); }
```

And correcting dslash_twisted_clover.cu (removed tuneKey, so we use the standard one, and added fillAux with different endings depending on the kernel called) I tried again, but I'm getting the same wrong results.

**Comment by mikeaclark** [comment]
Alex, is this problem still present in your latest development branch?  If so, I'll work on this tomorrow.

**Comment by AlexVaq** [comment]
I think it is, although I indeed solved a problem with the tuning. Let me try to check it today.

**Comment by AlexVaq** [comment]
Hi Mike, I can't easily test now on multiGPU. I have mostly access to my
own GPU in my linux box, but that's all, other GPUs are being used for
production. Single GPU works ok, but this is not telling us anything,
because the fused kernels are for multi GPU... so I don't know. I suppose
the problem still exists.

**Comment by mikeaclark** [comment]
Alex regarding your comment on needing multiple GPUs to test this: not sure if you were aware, but you can test multiple GPU code on a single GPU using the `--partition N` flag, where N is a 4 bit integer dictating which communications to switch on, with bits 0,1,2,3 used for dimensions x,y,z,t.  For example, 
```
--partition 1: x dimension communication only
--partition 6: y and z dimension communication
--partition 15: all dimensions
```
So this flag can be used to emulate multi-GPU performance using loopback communication.  It should be suitable for doing the testing required for this bug.

**Comment by AlexVaq** [comment]
No, I were not aware, and that's just... Wonderful!!! Tomorrow I'll give it a try (now it's too late).

**Comment by AlexVaq** [comment]
Just to make it clear. How should I run the code? Should I just add the --partition N flag? Or do I need to run two processes? How do I know comms are enabled and taken into account?

**Comment by mikeaclark** [comment]
You can run it on as many processes as you want, 1 or more; and you can mix and match the gridsize and the partition flag.  E.g.,
```
mpirun -np 2 ./dslash_test -tgridsize 2 --partition 15
```
Will run on two processes, splitting the time dimension between two GPUs, and enable loop-back communication in x,y,z.

**Comment by AlexVaq** [comment]
I found something interesting: if the volume is small enough (8^4), some partitioning works and gives the right result. I understood this as a memory problem, i.e. something is writting where it shouldn't.

In other cases, the inversion will diverge, unless tuning is off.

**Comment by mathiaswagner** [comment]
Does it depend on whether you start without an existing tune cache or use an already existing one?

On May 14, 2015, at 07:59, Alejandro Vaquero <notifications@github.com<mailto:notifications@github.com>> wrote:


I found something interesting: if the volume is small enough (8^4), some partitioning works and gives the right result. I understood this as a memory problem, i.e. something is writting where it shouldn't.

In other cases, the inversion will diverge, unless tuning is off.

**Comment by AlexVaq** [comment]
Not yet, I'll give it a try after lunch.

**Comment by azrael417** [comment]
Is there a Guard malloc or something like Valgrind for CUDA? If yes, then give it a try.



> Am 14.05.2015 um 05:34 schrieb Alejandro Vaquero <notifications@github.com>:
> 
> Not yet, I'll give it a try after lunch.
> 
> 
>

**Comment by azrael417** [comment]
Maybe something in the tunecache IO is broken. Do you get the same tunecache file when you perform tuning twice?



> Am 14.05.2015 um 05:34 schrieb Alejandro Vaquero <notifications@github.com>:
> 
> Not yet, I'll give it a try after lunch.
> 
> 
>

**Comment by mathiaswagner** [comment]
There is `cuda-memcheck`.

**Comment by AlexVaq** [comment]
I tuned some exterior kernels and then substitute type=exterior_[xyzt] for type=exterior_all manually in the cache file (otherwise I could’t tune anything). It worked in most of the cases. Once I got an error about not enough resources, but then I suppose the tuned parameters of the non-fused kernel were not good for the fused one.

**Comment by mikeaclark** [comment]
Ok, I'm looking at this now.
* Just running `dslash_test --dslash_type twisted_clover` fails the test.  Alex, is there a specific set of parameters I should use to make this work (to account for the fact that twisted-clover isn't implemented on the host)?
* Alex, did you push these patches, or are they outstanding?
```
    if (twist == QUDA_TWIST_GAMMA5_DIRECT) {
      writeAuxString("direct");
    } else {
      writeAuxString("inverse");
    }
```

```
TuneKey tuneKey() const { return TuneKey(in->VolString(), typeid(*this).name(), aux); }
```

**Comment by AlexVaq** [comment]
1. Yes, dslash_test will fail because clover is not implemented. here are several ways to overcome this issue and truly test that twisted_clover is properly working, but the best would be to implement clover on CPUs for testing. Workarounds:

	- Don’t create the clover term on the GPU, but feed it with a dummy clover term, like we do in wilson clover. Then you should compare to twisted-mass (I think this is the default) and you’ll get it, but this won’t assure you twisted-clover is really working, of course…

	- Create a non-trivial clover term for twisted-clover and wilson clover, and compare twisted-clover at mu=0.0 with wilson clover.

	- Get some external code (i.e. tmLQCD) to do the test.

If you just want not to get alarmed by the failed dslash-test (or for automated builds or whatever), just load the dummy clover term. But certainly we need for wilson clover and twisted-clover the CPU code that will allow us to truly test these operators.

2. These patches you mention were pushed by me, yes. Previously, I didn’t notice that the cases QUDA_TWIST_GAMMA5_DIRECT and QUDA_TWIST_GAMMA5_INVERSE were tuned with the same parameters, which I suppose is wrong. But I thought I pushed the patches only for the twisted-clover related operator (maybe I use the same enums than twisted-mass). Is this affecting twisted-mass?

**Comment by mikeaclark** [comment]
Ok, I'm confused: running 'dslash_test --partition 15 --dslash_type twisted-clover' with the fused boundary kernel enabled gives
```
ERROR: Twisted-clover dslash does not yet support a fused exterior dslash kernel (rank 0, host nvsocal1, dslash_twisted_clover.cu:162 in twistedCloverDslashCuda())
       last kernel called was (name=N4quda21TwistCloverGamma5CudaI6float4S1_EE,volume=12x24x24x24,aux=vol=165888,stride=165888,precision=4,TwistFlavour=-1)
```

**Comment by AlexVaq** [comment]
These lines were added long time ago in "void twistedCloverDslashCuda(…)" in dslash_twisted_clover.cu, when the fused kernels were not supported. If you want to use fused kernels you need to comment them:

    if (dslashPolicy ==  QUDA_FUSED_DSLASH || dslashPolicy == QUDA_FUSED_GPU_COMMS_DSLASH)
      errorQuda("Twisted-clover dslash does not yet support a fused exterior dslash kernel");

**Comment by mikeaclark** [comment]
Ok, I can't see the `writeAux` patch for dslash_quda.cu anywhere in the file.  This is supposed to be for TwistCloverGamma5Cuda?

**Comment by AlexVaq** [comment]
Exactly.

**Comment by mikeaclark** [comment]
But I thought you said you pushed this patch already?  What am I missing?

**Comment by AlexVaq** [comment]
I thought I pushed it. Isn’t it pushed?????? :S

**Comment by AlexVaq** [comment]
Ooooops, it’s pushed in the feature/dynamic_clover_inversion branch only. I suppose this is because of the new development model…
If I remember properly, this fix didn’t solve the problem, but let me try again.

**Comment by mikeaclark** [comment]
Let me make a quick pull request applying your patches, you can review and then merge into develop.

**Comment by mikeaclark** [comment]
Alex, on your comment above:
> And correcting dslash_twisted_clover.cu (removed tuneKey, so we use the standard one, and added  fillAux with different endings depending on the kernel called) I tried again, but I'm getting the same wrong results

Why do this?  Are these two things equivalent?

**Comment by AlexVaq** [comment]
I was using before my own version of tuneKey, but this might break compatibility in the future, so I decided to, let’s say, comply with QUDA standards, and use all functions inside QUDA. I think the final result should be the same, unless my code had some kind of memory overflow in the aux array (which, by the way, I think it had).

**Comment by mikeaclark** [comment]
Here's my pull request for fixing dslash_quda.cu: #241

Doing testing on the fused dslash doesn't give me any problems so far.  I'm wondering if this problem is coming from the dynamic clover inversion?

**Comment by AlexVaq** [comment]
Do you mean you think it was this tuning of TwistCloverDslashCuda that was spoiling everything? I tried right now with dynamic clover. I can tune, I don’t get errors, but the result is wrong. So it may be that the dynamic inversion is breaking the fused kernels.

I suppose this might postpone the merging of the feature/dynamic_clover_inversion branch into develop...

**Comment by mikeaclark** [comment]
What I'm saying at the moment, I think is:
* The addition of the direct/inverse label was definitely needed regardless of whether we were using fused kernels or not, since we could end up otherwise with invalid kernel launches.
* I don't see any problem with the develop branch as it stands (though I can't do a thorough test at the moment).

Can you test develop and your feature/dynamic_clover_inversion branch, but without dynamic clover?  Let see if this is what breaks the fused kernels.

**Comment by AlexVaq** [comment]
Ok, I have homework.

**Comment by mathiaswagner** [comment]
I've added a comment to put the pull request #221 on hold until we know more.

**Comment by mikeaclark** [comment]
Since fused twisted-clover seems to be working fine now, I've removed the code preventing this policy for  twisted clover (pull #264).

If there are problems with the dynamically inverted twisted-clover then we can deal that if it arises.
